rflx/pyrflx/typevalue.py

```diff
@@ -598,40 +635,62 @@ def set_field_with_length(field_name: str, field_length: int) -> Tuple[int, int]
                     )
             current_field_name = self._next_field(current_field_name)
 
-    def set(self, fld: str, value: Union[bytes, int, str, Sequence[TypeValue], Bitstring]) -> None:
-        if fld in self.accessible_fields:
-            field = self._fields[fld]
-            field.first = self._get_first(fld)
+    def set(
+        self, field_name: str, value: Union[bytes, int, str, Sequence[TypeValue], Bitstring]
+    ) -> None:
+        def set_refinement(fld: MessageValue.Field, fld_name: str) -> None:
+            if isinstance(fld.typeval, OpaqueValue):
+                for ref in self._refinements:
+                    if (
+                        ref.pdu.name == self.name
+                        and ref.field.name == fld_name
+                        and self._valid_refinement_condition(ref)
+                    ):
+                        fld.typeval.set_refinement(ref.sdu, self._refinements)
+
+        if field_name in self.accessible_fields:
+            field = self._fields[field_name]
+            field.first = self._get_first(field_name)
+            if isinstance(field.typeval, CompositeValue) and self._has_length(field_name):
+                field.typeval.set_expected_size(self._get_length(field_name))
+            set_refinement(field, field_name)
             if isinstance(value, Bitstring):
-                if isinstance(field.typeval, CompositeValue) and self._has_length(fld):
-                    field.typeval.set_expected_size(self._get_length(fld))
-                field.typeval.parse(value)
+                try:
+                    field.typeval.parse(value)
+                except (ValueError, KeyError) as e:
+                    raise ValueError(f"Error while setting value for field {field_name}: {e}")
             elif isinstance(value, field.typeval.accepted_type):
-                if isinstance(field.typeval, CompositeValue) and self._has_length(fld):
-                    field.typeval.set_expected_size(self._get_length(fld))
-                field.typeval.assign(value)
+                try:
+                    field.typeval.assign(value)
+                except (ValueError, KeyError) as e:
+                    raise ValueError(f"Error while setting value for field {field_name}: {e}")
```

Comment:
You could move the `try` block out of the `if` block.

Reply:
Fixed by b3ec99d